### PR TITLE
fix dbContext list

### LINKED ISSUE #791: data context list error
image

suggestion:

use `ActiveSolutionProjects`   instead of `StartupProjects`

https://github.com/ErikEJ/EFCorePowerTools/blob/master/src/GUI/EFCorePowerTools/Extensions/EnvDTEExtensions.cs#L63

use `Regex`   instead of `Split`
https://github.com/ErikEJ/EFCorePowerTools/blob/master/src/GUI/EFCorePowerTools/Handlers/ProcessLauncher.cs#L48
https://github.com/ErikEJ/EFCorePowerTools/blob/master/src/GUI/EFCorePowerTools/ViewModels/MigrationOptionsViewModel.cs#L232

### Further technical details
EF Core Power Tools version: 2.5.523

Database engine: SQL Server

Visual Studio version: Visual Studio 2019 16.8

### PR DESCRIPTION
**Pull request by Varorbc**
fix #791

### CODE REVIEW

File: src/GUI/efpt30.core/Program.cs [hunk 1]
```diff
@@ -10,6 +11,8 @@ static int Main(string[] args)
         {
             try
             {
+                Console.OutputEncoding = Encoding.UTF8;
```
Comment (ErikEJ) [inline src/GUI/efpt30.core/Program.cs:14]: I am guessing that this is actually the only real change required to fix this issue?

File: src/GUI/efpt30.core/EFCoreMigrationsBuilder.cs [hunk 2]
```diff
@@ -78,16 +78,8 @@ public class EfCoreMigrationsBuilder
 
             foreach (var type in types)
             {
-                try
-                {
-                    var dbContext = operations.CreateContext(type.Name);
-                    result.Add(new Tuple<string, string>(type.Name, GetMigrationStatus(dbContext)));
-                }
-                catch (InvalidOperationException ex)
```
Comment (ErikEJ) [inline src/GUI/efpt30.core/EFCoreMigrationsBuilder.cs:86]: Keep this in, it is in order to skip DbContext that cannot be created, and continue

File: src/GUI/EFCorePowerTools/Handlers/Compare/CompareHandler.cs [hunk 3]
```diff
@@ -156,7 +156,7 @@ private async Task<IEnumerable<string>> GetDbContextTypesAsync(string outputPath
                 throw new ArgumentException("Unable to collect DbContext information", nameof(processResult));
             }
 
-            if (processResult.StartsWith("Error:"))
+            if (processResult.Contains("Error:"))
```
Comment (ErikEJ) [inline src/GUI/EFCorePowerTools/Handlers/Compare/CompareHandler.cs:159]: Why this change?

File: src/GUI/efpt30.core/Program.cs (hunk 1 above)
Comment (Varorbc) [reply src/GUI/efpt30.core/Program.cs:14]: yes

File: src/GUI/efpt30.core/EFCoreMigrationsBuilder.cs (hunk 2 above)
Comment (Varorbc) [reply src/GUI/efpt30.core/EFCoreMigrationsBuilder.cs:86]: If the exception is handled and the user cannot know about it, why not tell the user?

File: src/GUI/EFCorePowerTools/Handlers/Compare/CompareHandler.cs (hunk 3 above)
Comment (Varorbc) [reply src/GUI/EFCorePowerTools/Handlers/Compare/CompareHandler.cs:159]: Like my case, the first output is not error

Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/Handlers/Compare/CompareHandler.cs:159]: I think I will update the tool with the encoding issue, then let's see if this is still the issue. Can you share the output with Error: that does not start with Error: ?

Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/Handlers/Compare/CompareHandler.cs:159]: I have updated the tool with just this change in the latest daily build.

File: src/GUI/efpt30.core/Program.cs (hunk 1 above)
Comment (ErikEJ) [reply src/GUI/efpt30.core/Program.cs:14]: I have fixed the encoding issue in the latest daily, please try that out, so we can see what else needs to be fixed. Please share the output if possible, rather than screenshots.

Comment (Varorbc) [reply src/GUI/efpt30.core/Program.cs:14]: 莫名其妙的问题\r\nDbContext:\r\nTestContext\r\nDebugView:\r\n\r\nDbContext:\r\nTest2Context\r\nDebugView:\r\n\r\n

File: src/GUI/EFCorePowerTools/Handlers/Compare/CompareHandler.cs (hunk 3 above)
Comment (Varorbc) [reply src/GUI/EFCorePowerTools/Handlers/Compare/CompareHandler.cs:159]: 莫名其妙的问题\r\n莫名其妙的问题\r\nError:\r\nSystem.InvalidOperationException: No database provider has been configured for this DbContext. A provider can be configured by overriding the 'DbContext.OnConfiguring' method or by using 'AddDbContext' on the application service provider. If 'AddDbContext' is used, then also ensure that your DbContext type accepts a DbContextOptions<TContext> object in its constructor and passes it to the base constructor for DbContext.\r\n   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.Initialize(IServiceProvider scopedProvider, IDbContextOptions contextOptions, DbContext context)\r\n   at Microsoft.EntityFrameworkCore.DbContext.get_InternalServiceProvider()\r\n   at Microsoft.EntityFrameworkCore.DbContext.Microsoft.EntityFrameworkCore.Infrastructure.IInfrastructure<System.IServiceProvider>.get_Instance()\r\n   at Microsoft.EntityFrameworkCore.Infrastructure.Internal.InfrastructureExtensions.GetService[TService](IInfrastructure`1 accessor)\r\n   at Microsoft.EntityFrameworkCore.
Infrastructure.AccessorExtensions.GetService[TService](IInfrastructure`1 accessor)\r\n   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(Func`1 factory)\r\n   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(String contextType)\r\n   at Modelling.EfCoreMigrationsBuilder.GetMigrationStatus(String outputPath, String startupOutputPath) in E:\\Projects\\EFCorePowerTools\\src\\GUI\\efpt30.core\\EFCoreMigrationsBuilder.cs:line 81\r\n   at Modelling.EfCoreMigrationsBuilder.GenerateMigrationStatusList(String outputPath, String startupOutputPath) in E:\\Projects\\EFCorePowerTools\\src\\GUI\\efpt30.core\\EFCoreMigrationsBuilder.cs:line 27\r\n   at Modelling.Program.Main(String[] args) in E:\\Projects\\EFCorePowerTools\\src\\GUI\\efpt30.core\\Program.cs:line 39\r\n

File: src/GUI/EFCorePowerTools/Handlers/ProcessLauncher.cs [hunk 4]
```diff
@@ -45,16 +46,13 @@ public Task<string> GetOutputAsync(string outputPath, GenerationType generationT
         {
             var result = new List<Tuple<string, string>>();
 
-            var contexts = modelInfo.Split(new[] { "DbContext:" + Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries);
-
-            foreach (var context in contexts)
+            var regex = new Regex("DbContext:\r\n(?<DbContext>[\\w]*)\r\nDebugView");
```
Comment (ErikEJ) [inline src/GUI/EFCorePowerTools/Handlers/ProcessLauncher.cs:49]: Please add a 5 second timeout

**Comment by ErikEJ** [comment]
Thanks for a great contribution, would be grateful if you can confirm the fix

**Comment by ErikEJ** [comment]
Just tested the RegEx, broke it for everyone else :-(

**Comment by Varorbc** [comment]
> Just tested the RegEx, broke it for everyone else :-(

What happened?

**Comment by ErikEJ** [comment]
DebugView content was left out - I did a much simpler fix: https://github.com/ErikEJ/EFCorePowerTools/commit/bc4ffb237b0b12d3c2f65c28610fa68c6d9c3030

**Comment by Varorbc** [comment]
Is it better to use json instead of string splicing for data transmission?

**Comment by ErikEJ** [comment]
Absolutely, that is what I do with the reverse engineering tool.

This is an old, naive implementation. 

But would json have solved the issue with the additional console output with chinese characters?

**Comment by Varorbc** [comment]
The console should be cleared `Console.Clear();`  before output to prevent unexpected strings

**Comment by ErikEJ** [comment]
Thanks, I will do that!

**Comment by ErikEJ** [comment]
I have just done that in the latest daily build

**Comment by Varorbc** [comment]
You should clear the console before output, not at the beginning, because the program has not output yet

**Comment by ErikEJ** [comment]
Learning something new! Thanks.

**Comment by ErikEJ** [comment]
Is the tool working for you now?
